### DISCUSSION TITLE
Addressing the Lack of Support for 'Store Notice' in Block Themes

### DISCUSSION THREAD

**Discussion by imanish003** [woocommerce/woocommerce#48526] [WooCommerce, Blocks, and Site Editing]
TLDR; Hiding “**Store notice”** in blocks themes isn’t straightforward or supported out of the box.

### "Notices" vs "Store Notice"

Before, I move forward, let's clarify the difference between "Notices" and "Store Notice". The former refers to the notices that are displayed using the "Notices" block. You may find more about it in pca54o-dl-p2 by @aljullu. Here is a screenshot of the "Notices" block:
![image](https://github.com/woocommerce/woocommerce/assets/16707866/10b8a6a7-6532-4103-84fe-55022adbb09f)

On the other hand, "Store Notice" is a store-wide notice that is available in classic themes. Here is a screenshot of the "Store Notice":
![image](https://github.com/woocommerce/woocommerce/assets/16707866/398bbace-7278-470a-a9ad-a6957f3504c2)

> [!IMPORTANT]
> This Github Discussion is focused on the "Store Notice" and not the "Notices" block.

### What's the issue?

There is [this issue](https://github.com/woocommerce/woocommerce/issues/42547) #42547, according to which it isn’t possible to hide “Store notice” in block themes. I commented on this issue on Nov 1, 2022 (See screenshot). Basically, I suggested to directly access the customizer settings using `/wp-admin/customize.php` path & hide the Store Notice. In the past few weeks, there have been multiple comments on this issue from HEs, mentioning that users are requesting this feature.
- There is [another comment](https://github.com/woocommerce/woocommerce/issues/42547#issuecomment-2140685269): "on block themes + narrow screens the site-wide notice is cut off"

In conclusion, "Store Notice" is supposed to be used with classic themes and it isn't supported out of the box in block themes. I believe this is a gap that needs to be addressed for block themes users.

### Solution? 

1. Creating a custom block?
  - We may consider developing a custom block for "Store Notice". This way, users can add this block to their pages/posts. Maybe we can use Block Hooks API to automatically add this block to every page/post. However, this is just an idea and needs further discussion.
2. Changing location of "Store Notice" setting? 
  - Another approach could be that we can move the "Store Notice" setting from the customizer to some other location. This way, users can still use this setting even with block themes. User won't need to manually go to the customizer to show/hide the "Store Notice". Maybe we can add this into "WooCommerce" settings in the admin panel. What are your thoughts on this?

### Feedback needed

I would like to hear your thoughts on this. Do you think we should address this gap? If yes, what could be the possible solutions? Are there any other alternatives that we can consider? Please share your feedback in the comments below. The goal of this Github Discussion is to gather feedback and find a solution to this problem 🙇

**Comment by jarekmorawski** [comment]
Thanks for starting this thread, @imanish003. I agree it's an important problem to solve! With blocks, the store notice itself (in the header, footer, or anywhere else) is no longer a component. It can be easily created with the core Paragraph or Group block and added to a template part that's displayed on every page.

That being said, I suspect there's nuance around extensibility and plugins. You can manage the notice's content in the Customizer and Woo's current onboarding UX, which means it has to be modifiable externally. In such a case, I'd argue we'd make it a custom block editable directly in the site editor.

**Reply by tjcafferkey** [reply]
I have a few questions:

* With classic themes the user can simply toggle this store notice on/off or edit the text on a global level in a single place. Would the equivalent in block themes be to add/remove the new custom block? Baring in mind we have two headers (store header & checkout header) so they would need to make sure they add/remove it from two header template parts at a minimum, maybe more if they create additional ones.
* To sync the store notice text globally (seeing as it _may_ appear in a number of templates/parts/places) we could use the block bindings API? I don't know enough about the API to say for sure if this is a good candidate but from my limited knowledge it seems like it. Its also currently compatible with the paragraph block.

**Reply by tjcafferkey** [reply]
Currently the store notice is extensible via the `woocommerce_demo_store` filter which allows extenders to modify the entire HTML of it. Currently all blocks have the `render_block` filter to do the same thing. So in short, by turning it into a block automatically provides the same level of extensibility in theory.

**Comment by Aljullu** [comment]
Thanks for starting this discussion, @imanish003! We definitely need to take another look at the _Customizer_ settings and make a plan on how to port them to block themes. I recently highlighted a similar issue we have with the WooCommerce Product Images settings (https://github.com/woocommerce/woocommerce/issues/28916#issuecomment-2162938417), which aren't accessible when using a block theme.

Regarding the Store Notice, I agree with @jarekmorawski's point of view that, in block themes, the Store Notice should be created with blocks. However, one of the issues we have is that somebody might activate a Store Notice using a classic theme, switch to a block theme and the notice will continue to be visible on the frontend with no way to hide it from the admin.

After some quick thinking, these are some of the ideas to handle those cases:

1. By default, hide store notices in block themes. To me, that seems the most desirable solution, as it clearly keeps the boundary between classic and block themes very clear: the store notice works in classic themes, but it doesn't in block themes. However, it has the big drawback that changing this now could be considered a breaking change, as some existing stores might be relying on it.
2. The other option would be, as you say, to expose a setting in the UI for block themes to modify or disable the Store Notice.
3. A third option would be to offer some kind of migration from the Store Notice setting into a blockified approach, but I think this adds a lot of complexity.

**Reply by jarekmorawski** [reply]
> A third option would be to offer some kind of migration from the Store Notice setting into a blockified approach, but I think this adds a lot of complexity.

I see your point, but it'd probably be the most scalable approach in the long term. It'd also be consistent with how we currently handle shortcode pages, like Cart and Checkout. We could:

- build a Store Notice block
- let classic theme users customize it using Customizer
- let block theme users customize it in the site editor as part of the Header template part
- display a placeholder when someone switches from classic theme to blocks, and let them migrate with a click of a button
- restore the classic notice element if someone switches back from blocks to a classic theme

**Comment by maxlaf** [comment]
I occasionally see users confused about this in support. I believe if we made these 2 changes it would eliminate the majority, if not all, of the confusion and resulting support requests we’ve seen about this topic:

* Disable the Store Notice by default on new installations using a block theme.
* Add a [Learn More] link in the default Store Notice message linking to the [public-facing documentation](https://woocommerce.com/document/woocommerce-customizer/) on the Customizer.

**Reply by tjcafferkey** [reply]
>Disable the Store Notice by default on new installations using a block theme.

From what I can see, [the setting defaults to `no` (disabled) by default](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/customizer/class-wc-shop-customizer.php#L356-L365) anyway on all sites. Do you mean disable it when a user switches themes from a classic theme to a block theme?

>Add a [Learn More] link in the default Store Notice message linking to the [public-facing documentation](https://woocommerce.com/document/woocommerce-customizer/) on the Customizer.

Users reading the store notice may think the link pertains to the messaging, when its actually about the feature (for the benefit of the merchant). We _could_ render that link only for logged in admins, though? If so would we need to consider making it obvious that regular users cannot see this text link

**Comment by ryanr14** [comment]
> From what I can see, [the setting defaults to no (disabled) by default](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/customizer/class-wc-shop-customizer.php#L356-L365) anyway on all sites. Do you mean disable it when a user switches themes from a classic theme to a block theme?

I'm glad to see this is disabled by default for new sites. I'd love to help figure out how we can best tidy this up, looking towards the blockified future. Like has been said by everyone the issue we run into is that the setting to control / change the Store Notice is hidden when switching to a block theme.

It seems like the smoothest "migration" option would be to detect the switch and disable the store notice and provide a store notice synced pattern in WooCommerce. I'd imagine that we insert an alert or notice within wp-admin somehow, at least if you are on a WooCommerce page within wp-admin.

It would say something like, "You've switched to a block theme, so we've disabled the previous Store Notice you had enabled. We now have a Store Notice block included in WooCommerce that you can add to your store using the Site Editor however you see fit."

I'd shared this in Slack, but it was pretty quick to recreate a store notice using the block editor.

https://github.com/user-attachments/assets/f6cc3b9b-c5ba-44cd-9371-b2f41e144bcf

This was my quick and dirty block code.

```
<!-- wp:group {"metadata":{"name":"Store Notice"},"className":"woocommerce-store-notice","style":{"color":{"background":"#d6b4e0","text":"#140e1f"},"elements":{"link":{"color":{"text":"#140e1f"}}}},"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group woocommerce-store-notice has-text-color has-background has-link-color" style="color:#140e1f;background-color:#d6b4e0"><!-- wp:group {"style":{"spacing":{"padding":{"right":"var:preset|spacing|50","left":"var:preset|spacing|50"}}},"layout":{"type":"flex","flexWrap":"nowrap","justifyContent":"left"}} -->
<div class="wp-block-group" style="padding-right:var(--wp--preset--spacing--50);padding-left:var(--wp--preset--spacing--50)"><!-- wp:paragraph {"style":{"layout":{"selfStretch":"fill","flexSize":null},"typography":{"fontStyle":"normal","fontWeight":"800","textTransform":"uppercase"}},"fontSize":"small"} -->
<p class="has-small-font-size" style="font-style:normal;font-weight:800;text-transform:uppercase">It's me, your store notice. Don't dismiss me!&nbsp;</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:spacer {"style":{"layout":{"selfStretch":"fill","flexSize":null}}} -->
<div aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->

<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap","justifyContent":"right"}} -->
<div class="wp-block-group"><!-- wp:buttons {"style":{"border":{"radius":"0px","width":"0px","style":"none"},"layout":{"selfStretch":"fit","flexSize":null}},"layout":{"type":"flex","justifyContent":"right","flexWrap":"nowrap"}} -->
<div class="wp-block-buttons" style="border-style:none;border-width:0px;border-radius:0px"><!-- wp:button {"width":100,"className":"woocommerce-store-notice__dismiss-link","style":{"border":{"radius":"0px"},"color":{"background":"#3c2861"}}} -->
<div class="wp-block-button has-custom-width wp-block-button__width-100 woocommerce-store-notice__dismiss-link"><a class="wp-block-button__link has-background wp-element-button" href="#" style="border-radius:0px;background-color:#3c2861">Dismiss</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

**Reply by Aljullu** [reply]
> It seems like the smoothest "migration" option would be to detect the switch and disable the store notice and provide a store notice synced pattern in WooCommerce.

Using a synced pattern is also a good idea. My main concern is that I'm not sure how easy it would be to make it work with [block hooks](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/), which is the system I had in mind to make the Store Notice to be automatically appended to the templates. :thinking: 

> I'd imagine that we insert an alert or notice within wp-admin somehow, at least if you are on a WooCommerce page within wp-admin.

About this, I'm not sure if an alert or notice is needed. I would expect any user switching themes to check their frontend to see how it looks. It should be clear whether the Store Notice has been migrated or not.

**Comment by Aljullu** [comment]
Heads-up that we published a GitHub Discussion with a proposal to reset _Customizer_ settings when switching from a classic theme to a block theme. If you have any thoughts or feedback, please let us know! 🙂 

* https://github.com/woocommerce/woocommerce/discussions/52840

**Comment by sverleis** [comment]
We have a customer affected by this issue, who needed to access via a dedicated link to the page: `/wp-admin/customize.php?return=%2Fwp-admin%2Fthemes.php`

9086052-zd-a8c
